### PR TITLE
Exclude out of bounds times

### PR DESCRIPTION
**Pull request by PaulRBerg**
## Description

Implements #1135. Use it by setting the `excludeOutOfBoundsTimes` prop on the `DatePicker` component. Example:

```jsx
<DatePicker
  selected={this.state.startDate}
  excludeOutOfBoundsTimes
  onChange={this.handleChange}
  minDate={new Date()}
  placeholderText="Click to select a date"
  showTimeSelect
/>
```

How it works:

- If you set a `minDate` or a `maxDate`, all times before that date's time will be deselected (if the active date matches the `minDate` or `maxDate`)
- If you switch dates and select an available time, but then come back to the `minDate` or `maxDate`, that time will not be highlighted. I accomplished this by going to the `renderTimes` function of the `Time` component and change the `base` to be `activeTime` instead of `new Date()` - I hope there won't be any unexpected side effects.

Caveat: I didn't have time to add tests, but I added examples in the documentation site.

### Min Date

Screen Shot 2019-04-21 at 16 28 25

### Max Date

Screen Shot 2019-04-21 at 16 28 44

### CODE REVIEW

**Comment by Akash187** [comment]
I guess "excludeOutOfBoundsTimes" is not implemented now in the official package. That's why it's not working for me.

**Comment by pcaplan-teladoc** [comment]
Great, I need this, when can it be merged and released?

**Comment by alfonmga** [comment]
We need to get this merged, it makes so much sense to have it.

File: src/date_utils.js [hunk 1]
```diff
@@ -379,6 +379,16 @@ export function isTimeInDisabledRange(time, { minTime, maxTime }) {
   return valid;
 }
 
+export function isTimeInExcludedRange(time, { minDate, maxDate }) {
```
Comment (CruseCtrl) [inline src/date_utils.js]: This function could be simplified to
```js
return isDate(minDate) && getDate(minDate) === getDate(time) && (getTime(minDate) > getTime(time) || getTime(maxDate) < getTime(time))
```

**Comment by rdrnt** [comment]
Any updates on this?

**Comment by danielcardoso5** [comment]
It would be really useful to have this merged. I can't block times selection for the current day without blocking them for every other day.

**Comment by jotak** [comment]
Hi @martijnrusschen , @CruseCtrl , 

Is there some reason to block this PR beside the only comment made by @CruseCtrl about a small simplification?
If that can help, I'm happy to take over this PR if there's more that needs to be worked on.

Thanks

**Comment by CruseCtrl** [comment]
I don't have any problem with it being merged

**Comment by thatguyjono94** [comment]
Please merge

**Comment by kylepeeler** [comment]
PLEASE merge it...

**Comment by khiner** [comment]
Also looking for this feature to enforce min/max datetime. Anything holding this up?

**Comment by PaulRBerg** [comment]
Hi @khiner, I am neither the author nor a maintainer of this library. I implemented the PR because I needed it for my own purposes at the time. @Hacker0x01 is the one to talk to.

**Comment by ajaxtown** [comment]
I made it work this way:

```jsx
const [startDate, setStartDate] = useState(new Date());

const IsSelectedDateToday = new Date().getDate() === startDate.getDate();

let minTimeHour = new Date().getHours();
if (!IsSelectedDateToday) minTimeHour = 0;

return (
  <DatePicker
      showTimeSelect
      selected={startDate}
      onChange={date => setStartDate(date)}
      minDate={new Date()}
      minTime={new Date(new Date().setHours(currentHour, 0, 0, 0))}
      maxTime={new Date(new Date().setHours(23, 59, 0, 0))}
      placeholderText="Select date and time"
   />
);
```

**Comment by ericvanular** [comment]
Please merge this...

Thanks @ajaxtown for the makeshift solution. Assuming you mean minTimeHour instead of currentHour in the minTime parameter. It will also set the min time as the hour you're currently in, so its still possible to set a time in the past.

**Comment by Jonofat** [comment]
> I made it work this way:
> 
> ```js-jsx
> const [startDate, setStartDate] = useState(new Date());
> 
> const isSelectedDateToday = new Date().getDate() === startDate.getDate();
> 
> let minTimeHour = new Date().getHours();
> if (!isSelectedDateToday) minTimeHour = 0;
> 
> return (
>   <DatePicker
>       showTimeSelect
>       selected={startDate}
>       onChange={date => setStartDate(date)}
>       minDate={new Date()}
>       minTime={new Date(new Date().setHours(currentHour, 0, 0, 0))}
>       maxTime={new Date(new Date().setHours(23, 59, 0, 0))}
>       placeholderText="Select date and time"
>    />
> );
> ```

This is extremely helpful, thank you. As mentioned in the comment below, the problem is that if you are within the same hour for example it still shows the time when it shouldn't. So, if it is 17:23 it is still allowing me to select 17:00 which it shouldn't. Any suggestions on how to fix that?

**Comment by Jonofat** [comment]
> Please merge this...
> 
> Thanks @ajaxtown for the makeshift solution. Assuming you mean minTimeHour instead of currentHour in the minTime parameter. It will also set the min time as the hour you're currently in, so its still possible to set a time in the past.

I have asked ajaxTown if they have a solution for the being able to set a time in the past but also asking you in case you had a solution?

**Comment by ajaxtown** [comment]
> So, if it is 17:23 it is still allowing me to select 17:00 which it shouldn't. Any suggestions on how to fix that?

@ericvanular @Jonofat you can manage the time this way:

```jsx
  const [startDate, setStartDate] = useState(new Date());

  const isSelectedDateToday = new Date().getDate() === startDate.getDate();

  let minTimeHour = new Date().getHours();
  if (!isSelectedDateToday) minTimeHour = 0;

  const date = new Date;
  const currentMins = date.getMinutes();
  const currentHour = date.getHours();

  return (
    <DatePicker
        showTimeSelect
        selected={startDate}
        onChange={date => setStartDate(date)}
        minDate={new Date()}
        minTime={new Date(new Date().setHours(hour, minutes, 0, 0))} // this is the change
        maxTime={new Date(new Date().setHours(23, 59, 0, 0))}
        placeholderText="Select date and time"
     />
);

```

**Comment by Jonofat** [comment]
@ajaxtown, thank you! This looks pretty good and is working for the current date. However if I choose tomorrow or any day in the future it still limits the time to what it is now eg: If it is 3:30pm it shows from 4pm. So I can't pick anything before 3:00pm for tomorrow which I should be able to.

I have had to change my code unfortunately before you posted your answer:

```
const DatePickerField = ({ ...props }) => {
  const { setFieldValue } = useFormikContext();
  const [field, meta] = useField(props);
  const [startDate, setStartDate] = useState(new Date());

  const isSelectedDateToday = new Date().getDate() === startDate.getDate();

  let minTimeHour = new Date().getHours();
  if (!isSelectedDateToday) minTimeHour = 8;

  const date = new Date();
  const currentMins = date.getMinutes();
  const currentHour = date.getHours();

  return (
    <span className="input-field">
      <DatePicker
        {...field}
        {...props}
        selected={(field.value && new Date(field.value)) || null}
        onChange={(val) => {
          setFieldValue(field.name, val);
        }}
        value={(field.value && new Date(field.value)) || null}
        minTime={new Date(new Date().setHours(currentHour, currentMins, 0, 0))}
        maxTime={new Date(new Date().setHours(17, 0, 0))}
        maxDate={addDays(new Date(), 3)}
        excludeTimes={excluded.map((exclude) => {
          if (field.value && field.value.getDay() === exclude.day) {
            return new Date(new Date().setHours(exclude.time, 0, 0));
          }
          return null;
        })}
        className="date-input"
      />
      <span className="input-error">
        {meta.error && meta.touched && meta.error}
      </span>
    </span>
  );
};

export default DatePickerField;
```

**Comment by ajaxtown** [comment]
@Jonofat ah yes, I missed that one out. Thanks for testing and letting me know. I have edited my answer to tackle that scenario as well.

**Comment by Jonofat** [comment]
@ajaxtown, thanks once again! I have edited my code above with your changes but it seems to still be doing the same thing ie: issue not resolved. I am trying to troubleshoot...

**Comment by Jonofat** [comment]
@ajaxtown  Ah, I spoke too soon! One last issue I think. If I have my max time is 5pm (ie: I should not be able to choose a time past 5pm on any given day ) and it is now 5:37 pm, it should show no times for today but it is showing everything from 00:00 to 23:00. It was working 100% when it was before 5pm today.

**Comment by ajaxtown** [comment]
@Jonofat can this be a timezone issue ?

**Comment by Jonofat** [comment]
@ajaxtown, I don't think it is a time zone issue ( I am working on my localhost). I have only allowed times between 8am and 5pm on Monday - Friday and 8am - 1pm on Saturday. But it is showing times for a whole 24 hours and ignoring my allowed times so I think it might be something else?

```
const excluded = [
  {
    time: "14",
    day: 6,
  },
  {
    time: "15",
    day: 6,
  },
  {
    time: "16",
    day: 6,
  },
  {
    time: "17",
    day: 6,
  },
];

  const isSelectedDateToday = new Date().getDate() === startDate.getDate();
  const isSelectedDateInFuture = +field.value > +new Date();

  const date = new Date();
  let currentMins = date.getMinutes();
  let currentHour = date.getHours();
  if (isSelectedDateInFuture) {
    currentHour = 8;
    currentMins = 0;
  }

minTime={new Date(new Date().setHours(currentHour, currentMins, 0, 0))}
maxTime={new Date(new Date().setHours(17, 0, 0))}

        excludeTimes={excluded.map((exclude) => {
          if (field.value && field.value.getDay() === exclude.day) {
            return new Date(new Date().setHours(exclude.time, 0, 0));
          }
          return null;
        })}
```

**Pull request closed without merging**